File: SORTING/InsertionSort/insertion_sort.py
from typing import List

def insertion_sort(arr: List[int], reverse: bool = False) -> List[int]:
    """
    Insertion Sort Algorithm.
    """
    n = len(arr)
    for i in range(1, n):
        j = i
        while j>0:
            if (not reverse and arr[j] < arr[j-1]) or (reverse and arr[j] > arr[j-1]):
                arr[j], arr[j-1] = arr[j-1], arr[j]
                j-=1
            else:
                break
        print(f"i: {i}, arr: {arr}")
    return arr


if __name__ == "__main__":
    test_arr = [32,23,12,43,65,33,1,34,5,45]
    sorted_arr = insertion_sort(test_arr, reverse=True)
    print(f"Sorted array: {sorted_arr}")

File: SORTING/InsertionSort/InsertionSort.java
package SORTING.InsertionSort;

import java.util.Arrays;

public class InsertionSort{
    public static int[] insertion_sort(int[] arr, boolean reverse){
        if(arr.length <=1){
            return arr;
        }
        int n = arr.length;
        for(int i = 1; i < n; i++){
            int j = i;
            while(j>0){
                if((!reverse && arr[j] < arr[j-1]) || (reverse && arr[j] > arr[j-1])){
                    int temp = arr[j];
                    arr[j] = arr[j-1];
                    arr[j-1] = temp;
                    j--;
                }
                else break;
            }
        }
        return arr;
    }

    public static int[] insertion_sort(int[] arr){
        return insertion_sort(arr, false);
    }

    public static void main(String[] args){
        int[] test_arr = {34,31,23,43,545,32,98,13,18};
        int[] sorted_arr = insertion_sort(test_arr, true);
        System.out.println("Sorted Array: " + Arrays.toString(sorted_arr));
        
    }
}